### DISCUSSION TITLE
why not use k8s service as upstream target instead of endpoints?

### DISCUSSION THREAD

**Discussion by I321065** [Kong/kubernetes-ingress-controller#4577] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

this is not bug or feature, I just want to know something about ingress controller design;
From ingress controller 2.10.x，I found the controller will parse k8s endpoints info and sync it to dataPanel to ask kong update
I have some confused about this , why not use service, for example service clusterIP, as the upstream target, for performance or other reason, please give me some tips if you can

**Comment by programmer04** [comment]
In brief (with possible use of mental shortcuts).

Using K8s Service `ClusterIP` directly as an upstream target would use the native K8s Service load balancing mechanism that is not controlled by Kong and adds some overhead (the actual difference varies in case what [CNI](https://github.com/containernetworking/cni) is used). Operating in Kong directly on a list of IPs allows bypassing the K8s Service mechanism and the DNS (otherwise, it'd be basically unnecessarily duplicated) and gives full control to Kong Gateway. To understand more read how networking works in Kubernetes
- [some fundamentals](https://opensource.com/article/22/6/kubernetes-networking-fundamentals)
- [official docs](https://kubernetes.io/docs/concepts/cluster-administration/networking/)
- to learn how networking works with the specific CNI let's check for instance EKS (AWS managed K8s)
    - [EKS workshops](https://www.eksworkshop.com/docs/networking/)
    - [some best practices](https://aws.github.io/aws-eks-best-practices/networking/index/)

> *If the above answers your question please click `Mark as answer` button to increase visibility for others.*
